### PR TITLE
Use new resource struct for matching / worker

### PR DESCRIPTION
**Pull request by wxing1292**
* Use new resource struct for matching initialization logic
* Use new resource struct for worker initialization logic
* Update worker scanner initialization logic
* Rewrite ringpop initialization logic to get rid of circular dependency

### CODE REVIEW

File: cmd/server/server.go [hunk 1]
```diff
@@ -211,9 +211,12 @@ func (s *server) startService() common.Daemon {
 	case historyService:
 		daemon = history.NewService(&params)
```
Comment (yux0) [inline cmd/server/server.go]: What about history and frontend

File: service/matching/handler.go [hunk 2]
```diff
@@ -37,38 +37,53 @@ import (
...
-	taskPersistence persistence.TaskManager
-	metadataMgr     persistence.MetadataManager
-	engine          Engine
-	config          *Config
-	metricsClient   metrics.Client
-	startWG         sync.WaitGroup
-	domainCache     cache.DomainCache
-	rateLimiter     quotas.Limiter
-	service.Service
+	resource.Resource
```
Comment (yux0) [inline service/matching/handler.go]: Is it necessary to keep the resource in the handler?

Comment (wxing1292) [reply service/matching/handler.go]: line 95 need to use it for rpc handler registration

File: service/matching/handler.go [hunk 3]
```diff
@@ -37,38 +37,53 @@ import (
...
+
+	taskMgr       persistence.TaskManager
```
Comment (meiliang86) [inline service/matching/handler.go]: You don't need to keep taskMgr, metadataMgr, domainCache, config here.

File: service/matching/service.go [hunk 4]
```diff
@@ -22,76 +22,71 @@ package matching
...
+	"github.com/uber/cadence/common/resource"
 	"github.com/uber/cadence/common/service"
 	"github.com/uber/cadence/common/service/dynamicconfig"
 )
 
 // Service represents the cadence-matching service
 type Service struct {
-	stopC  chan struct{}
-	params *service.BootstrapParams
+	resource.Resource
```
Comment (meiliang86) [inline service/matching/service.go]: is ServiceCommon/ServiceBase a better name?

File: service/matching/service.go [hunk 5]
```diff
@@ -22,76 +22,71 @@ package matching
...
+func NewService(
+	params *service.BootstrapParams,
+) (resource.Resource, error) {
+
+	serviceConfig := NewConfig(dynamicconfig.NewCollection(params.DynamicConfig, params.Logger))
+	serviceResource, err := resource.New(
+		params,
+		common.MatchingServiceName,
+		serviceConfig.ThrottledLogRPS,
+		func(
```
Comment (meiliang86) [inline service/matching/service.go]: Do you really need this func? Seems like all you are doing is calling persistenceBean.GetVisibilityManager()

Comment (wxing1292) [reply service/matching/service.go]: frontend / history will have some overrides

File: service/matching/service.go (hunk 4 above)
Comment (wxing1292) [reply service/matching/service.go]: i will change this resource thing to service after the moving everything, then delete service, then renaming it to service

File: service/matching/handler.go (hunk 3 above)
Comment (wxing1292) [reply service/matching/handler.go]: done

File: service/matching/handler.go (hunk 2 above)
Comment (wxing1292) [reply service/matching/handler.go]: making every component (high level) having access to resource is the idea here.

**Pull request merged**